include/ring_buffer.h

```diff
@@ -25,8 +25,8 @@ extern "C" {
  * @brief A structure to represent a ring buffer
  */
 struct ring_buf {
-	u32_t head;	 /**< Index in buf for the head element */
-	u32_t tail;	 /**< Index in buf for the tail element */
+	volatile u32_t head;	 /**< Index in buf for the head element */
+	volatile u32_t tail;	 /**< Index in buf for the tail element */
```

Comment:
I think this is the wrong spot.   There are plenty of uses for a ring buffer that don't involve contending with SMP CPUs.  Instead, I think you probably want to tag the pointer to the struct volatile in the app instead.

Reply:
Ok, Thanks

Reply:
> One nitpick about volatile, one spot in the SPIN_VALIDATE predication that definitely needs to be changed.
I have done the code change, but I should git push it after #14863 has been merged, otherwise it cann't pass the shippable. So please review #14863 first! Thanks!

Reply:
Hate to stick my nose in here, but is `volatile` even the right thing here?  If the issue is that other CPUs might change the value, then I'm not convinced that we don't also need explicit storage fences.  Probably not a problem on x86, but on other arches.

Reply:
On architectures where it matters, the fence will be provided by the synchronization primitive where we are using one, which we aren't here.  It's true that IPM is sort of a special case, as it was intended for a asymmetric multiprocessing architecture (Quark SE -- one x86 and one ARC on the same memory bus) with no cache.  But that's not really something we can fix with this patch.